homeassistant/components/stream/hls.py

```diff
@@ -63,56 +77,259 @@ class HlsPlaylistView(StreamView):
+            len(segment.parts),
+            segment.complete,
+        )
+
+    @classmethod
+    def render(cls, track, ll_hls):
+        """Render HLS playlist file."""
+        # NUM_PLAYLIST_SEGMENTS+1 because most recent is probably not yet complete
+        segments = list(track.get_segment())[-(NUM_PLAYLIST_SEGMENTS + 1) :]
+
+        # Create a copy of the last segment with the parts frozen for rendering playlist
+        segments[-1] = copy(segments[-1])
+        segments[-1].parts = segments[-1].parts.copy()
 
-        if not segments:
-            return []
+        # Remove first segment if the last segment is actually complete
+        if segments[-1].complete:
+            segments = segments[-NUM_PLAYLIST_SEGMENTS:]
+
+        # The track may have been updated since we froze it, but this should be good enough
+        target_duration = track.target_duration
 
         playlist = [
+            "#EXTM3U",
+            "#EXT-X-VERSION:6",
+            "#EXT-X-INDEPENDENT-SEGMENTS",
+            '#EXT-X-MAP:URI="init.mp4"',
+            f"#EXT-X-TARGETDURATION:{target_duration}",
             f"#EXT-X-MEDIA-SEQUENCE:{segments[0].sequence}",
             f"#EXT-X-DISCONTINUITY-SEQUENCE:{segments[0].stream_id}",
+            "#EXT-X-PROGRAM-DATE-TIME:"
+            + segments[0].start_time.strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3]
+            + "Z",
         ]
 
-        last_stream_id = segments[0].stream_id
-        for segment in segments:
-            if last_stream_id != segment.stream_id:
-                playlist.append("#EXT-X-DISCONTINUITY")
+        if ll_hls:
+            part_duration = float(
+                max(
+                    (p.duration for s in segments for p in s.parts),
+                    default=PART_DURATION,
+                )
+            )
             playlist.extend(
                 [
-                    f"#EXTINF:{float(segment.duration):.04f},",
-                    f"./segment/{segment.sequence}.m4s",
+                    f"#EXT-X-PART-INF:PART-TARGET={part_duration:.3f}",
+                    f"#EXT-X-SERVER-CONTROL:CAN-BLOCK-RELOAD=YES,PART-HOLD-BACK={2*part_duration:.3f}",
+                    f"#EXT-X-START:TIME-OFFSET=-{3*part_duration:.3f},PRECISE=YES",
                 ]
             )
+        else:
+            # Since our window doesn't have many segments, we don't want to start
+            # at the beginning or we risk a behind live window exception in exoplayer.
+            # EXT-X-START is not supposed to be within 3 target durations of the end,
+            # but this seems ok
+            playlist.append(
+                f"#EXT-X-START:TIME-OFFSET=-{1.5 * target_duration:.3f},PRECISE=YES"
+            )
+
+        last_stream_id = segments[0].stream_id
+
+        # Add playlist sections for completed segments
+        # Enumeration used to only include EXT-X-PART data for last 3 segments.
+        # The RFC seems to suggest removing parts after 3 full segments, but Apple's
```

Comment:
Apple's examples near the bottom of [this page](https://developer.apple.com/documentation/http_live_streaming/enabling_low-latency_hls) show parts for the in progress segment + the 2 prior segments.
Meanwhile, the example in the RFC [here](https://tools.ietf.org/html/draft-pantos-hls-rfc8216bis-09#section-8.11) shows parts for the in progress segment + the 3 prior segments.
Currently we are just following the method used in Apple's examples, and this seems to work fine.